cohortextractor/main.py

```diff
@@ -201,3 +224,9 @@ def write_output(results, output_file):
             else:
                 assert fields == headers, f"Expected fields {headers}, but got {fields}"
             writer.writerow(entry.values())
+
+
+def write_validation_output(results, output_file):
+    with output_file.open(mode="w") as f:
+        for entry in results:
+            f.write(f"{str(entry)}\n")
```

Comment:
This might not matter for our purposes at the moment, but this won't produce the actual SQL that's going to be executed as it will just output the SQL string with placeholders and won't interpolate the values into the placeholders.

This occurred to me because @bloodearnest was just asking how to generate the full SQL from a study and I was going to point him here.

If we _do_ want to support this at some point then, just for reference, here's the function we used to have to do literal interpolation:
https://github.com/opensafely-core/cohort-extractor-v2/blob/f9b554d8c102a2391327d270e157d69d651a8e83/cohortextractor/query_engines/mssql.py#L622-L634

And we may also need to teach types in the Spark dialect how to render themselves as literals. For reference, this is how we do it for the custom MSSQL date types:
https://github.com/opensafely-core/cohort-extractor-v2/blob/f6ad508948b75685deebeb53a8384e719088853b/cohortextractor/query_engines/mssql.py#L36-L45

Reply:
No, I realised it wouldn't produce the actual SQL, but I wasn't sure how much exporting the SQL was useful for researchers anyway, and I thought the placeholder SQL string might still be useful for us for debugging.  Will that literal interpolation work if it isn't actually connecting to a database (it won't be for local validation)?

Reply:
Yeah, you're right that it's not needed by the researchers at the moment and the current output will still be useful for us. I think further down the line it probably will be useful to be able to get the exact SQL as executed so I just wanted to make a record of the pieces we'd need in place, but this definitely wasn't something that needs doing now.

The literal interpolation does run without a database connection, but where the dialect has features that depend on the specific database version (as with `supports_multivalues_insert` above) then we have to manually set those if we don't have a real database to check support against.